### PR TITLE
Minor Refactoring to LCP

### PR DESCRIPTION
**Pull request by franklinselva**

### CODE REVIEW

File: planners/src/solver.rs [hunk 1]
```diff
@@ -0,0 +1,272 @@
+use crate::encode::{encode, populate_with_task_network, populate_with_template_instances};
+use crate::fmt::{format_hddl_plan, format_partial_plan, format_pddl_plan};
+use crate::forward_search::ForwardSearcher;
+use crate::solver::Strat::{Activity, Forward};
+use crate::Solver;
+
+use anyhow::{Context, Result};
+use aries_model::state::Domains;
```
Comment (franklinselva) [inline planners/src/solver.rs:8]: Possibility of missing `state.rs` in `master` branch.

https://github.com/plaans/aries/tree/master/model/src

Comment (franklinselva) [inline planners/src/solver.rs:8]: Possibility of missing `state.rs` in `master` branch.

https://github.com/plaans/aries/tree/master/model/src

File: planners/src/solver.rs [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+    pub end: Instant,
+    pub htn_mode: bool,
+}
+
+impl Planner {
+    pub fn new(option: Opt) -> Self {
+        Self {
+            option,
+            problem: None,
+            plan: None,
```
Comment (franklinselva) [inline planners/src/solver.rs:69]: `plan: None` -> Only builds in local machines

**Comment by arbimo** [comment]
All pull requests should be based on the master branch. Could you rebase this ?

Also, to avoid any unhandled lints, you should make sure that you have the latest rust version (`rustup update`).

**Comment by franklinselva** [comment]
> All pull requests should be based on the master branch. Could you rebase this ?
> 
> Also, to avoid any unhandled lints, you should make sure that you have the latest rust version (`rustup update`).

I will rebase the changes from `temporal-pddl` branch.

The RUST has been updated and the version is `1.59.0`. I will check for the issue

**Comment by franklinselva** [comment]
Adding PR #24 to address rebase issue.

File: planners/src/solver.rs [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+use aries_planning::chronicles::Problem;
+use aries_planning::chronicles::*;
+
+use aries_solver::parallel_solver::ParSolver;
+use aries_tnet::theory::{StnConfig, StnTheory, TheoryPropagationLevel};
+
+/// An automated planner for PDDL and HDDL problems.
+#[derive(Debug, Default, Clone, StructOpt)]
+#[structopt(name = "grpc", rename_all = "kebab-case")]
+pub struct Opt {
```
Comment (arbimo) [inline planners/src/solver.rs:28]: This should stay in the lcp.rs file.

File: planners/src/solver.rs [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+    /// partial plan will be displayed.
+    #[structopt(long = "no-search")]
+    no_search: bool,
+    /// If provided, the solver will only run the specified strategy instead of default set of strategies.
+    /// When repeated, several strategies will be run in parallel.
+    #[structopt(long = "strategy", short = "s")]
+    strategies: Vec<Strat>,
+}
+
+pub struct Planner {
```
Comment (arbimo) [inline planners/src/solver.rs:55]: There is no reason for the Planner to be stateful. It is probably better to only have a set of free-standing functions with well identified parameters. So the `Planner` struct should be removed and the method take their needed parameters explicitly.

File: planners/src/solver.rs [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+            end: Instant::now(),
+            htn_mode: false,
+        }
+    }
+
+    pub fn get_answer(&self) -> Option<Arc<Domains>> {
+        self.plan.clone()
+    }
+
+    pub fn solve(&mut self, mut spec: Problem, opt: &Opt) -> Result<()> {
```
Comment (arbimo) [inline planners/src/solver.rs:80]: Make a function: remove the `self` and pass needed parameters explicitly (breaking out the `Opt` into individual parameters). It should not have access to any uneeded information (e.g. the problem filename).

Could be renamed `deepening_solve`.

File: planners/src/solver.rs [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+                format_pddl_plan(
+                    &self
+                        .problem
+                        .clone()
+                        .with_context(|| "Unable to format PDDL problem. Formatting failed".to_string())?,
+                    x,
+                )?
+            };
+            println!("\n**** Plan ****\n\n {}", plan);
+            if let Some(plan_out_file) = self.option.plan_out_file.clone() {
```
Comment (arbimo) [inline planners/src/solver.rs:170]: Writing out the plan to a file should stay in the lcp.rs file.

File: planners/src/solver.rs [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "1" | "act" | "activity" => Ok(Activity),
+            "2" | "fwd" | "forward" => Ok(Forward),
+            _ => Err(format!("Unknown search strategy: {}", s)),
+        }
+    }
+}
+
+fn _solve(pb: &FiniteProblem, opt: &Opt, htn_mode: bool) -> Option<std::sync::Arc<SavedAssignment>> {
```
Comment (arbimo) [inline planners/src/solver.rs:231]: Rename `solve_finite_problem`. Split the `Opt` into several args.

File: planners/src/solver.rs [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+            let plan = if self.htn_mode {
+                format!(
+                    "\n**** Decomposition ****\n\n\
+                        {}\n\n\
+                        **** Plan ****\n\n\
+                        {}",
+                    format_hddl_plan(
+                        &self
+                            .problem
+                            .clone()
```
Comment (arbimo) [inline planners/src/solver.rs:148]: I don't believe the clone to be useful there.

File: planners/src/solver.rs [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+                            .with_context(|| "Unable to format HDDL problem. Formatting failed".to_string())?,
+                        x
+                    )?,
+                    format_pddl_plan(
+                        &self
+                            .problem
+                            .clone()
```
Comment (arbimo) [inline planners/src/solver.rs:155]: idem

File: planners/src/solver.rs [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+                            .with_context(|| "Unable to format PDDL problem. Formatting failed".to_string())?,
+                        x
+                    )?
+                )
+            } else {
+                format_pddl_plan(
+                    &self
+                        .problem
+                        .clone()
```
Comment (arbimo) [inline planners/src/solver.rs:164]: idem

File: planners/src/solver.rs [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+            }
+            if result.is_some() {
+                self.problem = Some(pb.clone());
+                break;
+            }
+        }
+
+        self.end = Instant::now();
+        self.plan = result;
+        Ok(())
```
Comment (arbimo) [inline planners/src/solver.rs:133]: The plan should be returned (actually necessary without the `Planner` instance).

File: planners/src/solver.rs [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+    }
+
+    pub fn format_plan(&self, plan: &Option<Arc<Domains>>) -> Result<()> {
```
Comment (arbimo) [inline planners/src/solver.rs:136]: This performs many side effects and should probably remain in `lcp.rs`. The logic is actually quite simple and different `main` implementation are likely to want different behaviors anyway.

In general, the method name should have been `print_plan` as it does more than just formatting.

**Review by arbimo — CHANGES_REQUESTED** [review]
@franklinselva I corrected the bug you had. It was caused by the reuse of the base problem which resulted in an exponential growth in its size.

I made some comment on the existing implementation.
The overall idea is to simplify the implementation to only have a couple of state-less functions with explicit parameters (instead of the `Opt` that is specific to LCP and is bound to grow over time).

**Pull request merged**